### ISSUE TITLE
Make compatible with Horus transmitters

### ISSUE THREAD

**Issue by teckel12** [teckel12/LuaTelemetry#55]
Make script compatible with Horus transmitters
* [x] Write replacement for lcd.getLastPos() to work on Taranis and Horus (Horus-only view resolves this)
* [x] Find solution for Horus' lack of `ERASE` and `FORCED` flags (Horus-only view resolves this)
* [x] Horus-only view: Pilot panel
* [x] Horus-only view: Radar panel
* [x] Horus-only view: Flight data panel
* [x] Config menu adjustments to work on Horus
* [x]  Allow script to work as a widget
* [x] Create larger Horus-specific icons
* [x] Provision for multilingual support for data field labels
* [x] Update documentation on how to use with Horus
* [ ] Optimize code to increase speed/reduce memory
* [x] Certain inversions don't display correctly on attitude display (also happens on Taranis?)
* [x] Remove/recolor background filled rectangles on config menu

**Comment by teckel12** [comment]
Now that 1.4.0 adds multiple views, extend that to accommodate Horus with it's own custom view.

**Comment by fmartinezo** [comment]
I understand little programming in c ++ and lua, but I'm about to get a 10 horus and I'm interested in the topic

**Comment by teckel12** [comment]
@fmartinezo Stock, a Horus doesn't support Lua. I believe you need to switch to OpenTX to "kinda" work with Lua. I also believe you need to manually start each Lua script each flight, not automatic like with Taranis.

**Comment by fmartinezo** [comment]
Yes, It will be a mandatory request to run lua on horus equal as you need some requirements in order to run your lua in taranis

**Comment by fmartinezo** [comment]
I have been thinking about telemetry for the Horus and I suppose that these work in a similar way to the taranis, at least with openTX.
In the taranis we have 4 screens to show telemetry data, and even by special functions we can make taranis say its value, but there is a problem with the sensors of TMP1 and TMP2 that are formed by multiple values and that except for lua I do not see how to split them.

Reading reading, I have found the theme of the widgets but there I am lost.

It would not be possible to make a specific lua for the horus, which at the moment only represents values, without graphics, and that is able to speak as the current lua does? I say it for the codes that you mention of ERASE and FORCED

Also I've seen this that can be interesting as a starting point, since it's open source https://github.com/yaapu/FrskyTelemetryScript it is for mavlink, but I guess it would be reprogrammable for inav

**Comment by teckel12** [comment]
@fmartinezo Here you go, this is as far as I could take it: https://github.com/iNavFlight/LuaTelemetry/archive/horus.zip

I can only run it as a one-time script, I can't figure out how to use the flippin' transmitter to do it as a widget.  Maybe because I'm trying to do it with the simulator, but nothing seems to work the way any tutorial describes.

In any case, while it "works" the display is (as expected) totally wrong.  But, I can't get further than this because it doesn't seem to see any of my telemetry sensors.  Nothing works.

This is my frustration with this transmitter.  It's half-baked, and maybe half of that in the simulator.  Someone needs to take the above branch and give instructions as to how to actually get it working.  For example, what exact button strokes are required to make a widget?  How the heck do you get the telemetry sensors working in the script?

Once someone with a Horus transmitter can answer these questions and be willing to help can this go further. If you want Horus support, you're going to need to do some basic work on your end first as I'm stuck at this point.

**Comment by teckel12** [comment]
I figured out the telemetry. On Horus I was running as a one-time script, which requires me to call the background task (which is where the telemetry is gathered).

Anyway, that's sorted out so now I just need to make a Horus view and make the config screen formatted better on the Horus.

I'd still like to look into running Lua Telemetry from a widget, so any help in that area would be appreciated.

**Comment by fmartinezo** [comment]
hi Tim
I tried to execute the lua that you sent me from companion but I can not get it to work decently.
I have internally changed the value of the "path" so that it takes the files from the "widgets" directory but I can not get it out. I have run it from companion looking for it in the sd, but this is a single execution and what comes out is quite deficient.
There is a page https://github.com/yaapu/FrskyTelemetryScript/wiki/How-to-run-the-Yaapu-script-on-X10-and-X12-as-a-Widget that explains how to put it to run as a widget and therefore be permanent and load while the model is turned on, but I have not achieved anything at the moment because at the point where it says that it puts the name of the lua to me does not appear iNav. All my tests are from companion since at the moment I do not have the transmitter. However, I'm still looking

**Comment by teckel12** [comment]
It's only running as a one-time script currently, and by running its not doing anything other than showing RSSI, as the view hasn't been created. It can't use a view from Taranis as the screen and font sizes make it unusable, it has to be totally rebuilt.

I know how to start a widget, the question I have is how to build the script to work as a widget. But this may require someone knowledgeable with Horus widgets and Lua Scripts. Not that I can't eventually figure it out, just doesn't make much sense at first glance.

**Comment by teckel12** [comment]
Attitude view is looking good now, a few more things to add and I can move on to the map view.  After that I'll work on displaying the rest of the data.

**Comment by fmartinezo** [comment]
fantastic, those are good news

**Comment by teckel12** [comment]
Sometime tomorrow I'll port a few more features to the Horus view then post a pic of what it's looking like. Better to address any issues early.

**Comment by teckel12** [comment]
![screenshot_x10_18-11-19_08-28-12](https://user-images.githubusercontent.com/2592128/48710137-6f242000-ebd5-11e8-8c19-9a94a241661b.png)

Only the title and glass cockpit display are finished, but it gives you an idea of where I'm heading.  The green area will be the radar view, between the glass cockpit and radar will be the variometer, and the blue area will be the flight data.  Haven't got to the layout of the flight data, but it will probably be similar to the X9D's radar view.

**Comment by fmartinezo** [comment]
It looks very nice. Very very very good job

**Comment by teckel12** [comment]
@fmartinezo Latest progress...
![horus](https://user-images.githubusercontent.com/2592128/48926443-22be3600-ee9c-11e8-93db-a18e588095c7.png)

**Comment by fmartinezo** [comment]
@teckel12 Wow, it looks better than spected. Several questions...
Battery is per cell or is total battery?
What are 44.1 and 374?
Were are GPS coodenates?

In a few days my horus will come home. If you wish, I will notify you so that you can provide me with the code to do tests.

**Comment by teckel12** [comment]
@fmartinezo Battery can be displayed per cell or total (config option). The 44.1/374 are speed/altitude, which is a common glass pilot display. Speed and altitude will also be displayed in the blank area at the bottom/center (just haven't programmed that yet). Same is true with the GPS data including coordinates, that will be on the bottom right.

See the pilot and radar view pictures for the X9D to get an idea of what it will look like and what the values mean.  There's also a Wiki which describes the views which will make everything on the Horus make sense.

**Comment by teckel12** [comment]
For the most part, the Horus view is complete.  There needs to be some polish but the data is all there.  I need to make the config menu format correctly on Horus and then I'd like some beta testers.  Config menu shouldn't take too long, so maybe sometime this week it will be ready for testing.

![horus](https://user-images.githubusercontent.com/2592128/48984919-e5da8500-f0cf-11e8-8c16-962080e0ea9e.png)

![horus2](https://user-images.githubusercontent.com/2592128/48984921-ea06a280-f0cf-11e8-91ab-c18694dec6c9.png)

**Comment by fmartinezo** [comment]
I have already received my brand new horus x10s so when you have ready the telemetry I can start the tests.
By the way is it going to be a permanent widget or one-time script?

**Comment by teckel12** [comment]
@fmartinezo The plan is both a widget and one-time script. Currently, it's only a one-time script.

I've added a GPS signal strength indicator and finishing up larger Horus-specific icons. Next, I'll rework the config menu, which shouldn't be too challenging. On-schedule to test later this week.

**Comment by fmartinezo** [comment]
Great. I'm waiting for it. Ready to check

**Comment by fmartinezo** [comment]
ok, I'll start with the tests and I'll tell you .....
What luck with the horus and frsky, I also want to send me another ..... :)
At the moment I've found something silly but it's going to make you change the code. There is a sound whose wav file has more than six letters which is the maximum allowed in horus for the names is the "hedhlda.wav"

**Comment by teckel12** [comment]
@fmartinezo I believe the file name limits have to do with using the sound files in other parts of the transmitter.  In Lua scripts, there's not these limitations.  The Q7X also has a 6 character limit for audio files, but this audio file works fine on the Q7X in Lua Telemetry.  Also, testing in the Horus simulator plays the audio file just fine.

You'd need to confirm it actually doesn't work on the transmitter, as everything I've experienced says that the 6 character limit doesn't pertain to Lua scripts.

**Comment by fmartinezo** [comment]
@teckel12 Hi. I started with the tests and I had to reset the horus twice. I tell you the first one was when loading telemetry for the first time, since internally it generates a "luac". the second, for the same reason when I tried to show the menu. After this and already generated the corresponding luac, it works without blockages. I attached the two screens in scoop in my horus
![img_20181128_205315_640x480](https://user-images.githubusercontent.com/38476783/49178897-4c220a80-f351-11e8-94a7-b45f55fe72ea.jpg)
![img_20181128_205635_640x480](https://user-images.githubusercontent.com/38476783/49178903-504e2800-f351-11e8-817d-94bac5ea2c23.jpg)

**Comment by fmartinezo** [comment]
The tests I just did this afternoon / night in my house, so he has not fixed satellites. Tomorrow if I can I will go out to the street to do some more test.
congratulations for your work

**Comment by teckel12** [comment]
@fmartinezo When you say you had to reset the Horus twice, what did you need to do?  Was it that you had to start the script again or the the entire Horus crashed or need to be rebooted?  Also, do you know what build options you have selected for the OpenTX firmware you're running?  These are the check boxes you select in settings in the OpenTX Companion when downloading the OpenTX firmware.

I made a few mostly cosmetic changes to the `horus` branch.  The two main changes you'll see is a larger font size in the config menu and the white box between the attitude and map is gone (when the variometer is turned off).  Use the same process to upgrade to the latest build (the steps 1-4 shown above).

I noted you're using the Spanish language translations.  There's a few new labels on the Horus display that are longer than on Taranis.  I believe the following could be translated and incorporated into the language files which would be useful for non-English speakers:

```
Fuel
Battery
Altitude
Distance
Current
```

A Spanish translation for the above would be helpful.

Finally, there's a testing-only number in the title between the model name and the TX battery gauge.  It shows the frames per second that the display is updating.  When not armed, you can long-press the enter button which will reset the script as well as the FPS number.  In the above screen shots, it shows you're running at about 5 frames per second.  I'd rather see more like 10 or higher, but there's a lot going on so it's kind of expected that it's not very high.  Also, telemetry doesn't get updated very frequently anyway, so it's not too much of a concern.  However, moving forward I would like to try and optimize the code a bit to make it faster.  So keep an eye on the FPS to see if we're going in the correct direction.

**Comment by fmartinezo** [comment]
The tanslation will be
Fuel = no translation because it is hard to explain. This is the estimated percent remaining for battery
Battery = Batería
Altitude = Altitud
Distance = Distancia
Current = Consumo

**Comment by fmartinezo** [comment]
What I meant by restarting the horus is that it was blocked and did not recognize any system or rtl key. I had to press the reset button to turn off the horus.
The version that comes down from opentx firmware was in Spanish, luac and luac marked
![imagen](https://user-images.githubusercontent.com/38476783/49187661-8519a980-f368-11e8-8dd2-d56554cb0d40.png)
![imagen](https://user-images.githubusercontent.com/38476783/49187679-95ca1f80-f368-11e8-8903-28f1d30d6ccb.png)

**Comment by fmartinezo** [comment]
Clarifying more about the restart I had to insert a pin on the emergency off button that has the horus next to the slot of the sd card as it did not accept to disconnect with the switch and did not respond to system buttons, etc.

**Comment by fmartinezo** [comment]
For what you say, I understand that you have made changes in the software and that I have to download it again. Is it like that?

**Comment by fmartinezo** [comment]
I think I have misunderstood you in relation to the translation. Do you need literals or wav files?

**Comment by teckel12** [comment]
@fmartinezo Yes, just download and install it again. It includes some cosmetic changes and it's possible it won't lock up again when you run it the first time. This is the kind of problem having my own transmitter will allow me to diagnose easier.

As for the Spanish translations, that's perfect, it's for the literal screen view.

I also played around with running the script as a widget. It's not really designed to have a single code base for Taranis telemetry script and a Horus widget. But, I believe I can make things work. I got the screen to display without any errors, so I feel confident it's possible.

**Comment by fmartinezo** [comment]
@teckel12 Hi Tim. I just downloaded and installed the latest version. It has not been blocked the first time I have executed it and that I have started from scratch. I deleted everything and installed everything again.
I managed to block it and I tell you how. You turn on the transmitter only, you execute the script and it crashes. Then you turn on the receiver but it is still blocked. It gives the impression that when not receiving telemetry values it is blocked.
I have seen the fps in 5.2 but I have also seen a value around 7. It was in a screen change to see the options menu and return to the telemetry screen.
I'm going out on the street to set the gps and see how it goes, in addition to listening to the voices.
I'll tell you later

**Comment by fmartinezo** [comment]
I just went up the street and the tests have been satisfactory. The sounds are correct and the script goes well. There is only the litmus test (definitive test) that will be in flight since it is forbidden to fly in the city. Here you have two photos. Do not miss out on the border because I have deleted the gps coordinates
![img_20181129_132450_1280x720](https://user-images.githubusercontent.com/38476783/49222796-24788400-f3dd-11e8-8885-5f53cdc4bbe6.jpg)
![img_20181129_132501_1280x720](https://user-images.githubusercontent.com/38476783/49222800-27737480-f3dd-11e8-96bb-740a6b66f7d0.jpg)

**Comment by teckel12** [comment]
@fmartinezo A few more updates to the Horus branch today.  Nothing should look different, but I changed how the screen is drawn and I'd be interested to see if the frames per second count is better, worse, or indifferent.

**Comment by fmartinezo** [comment]
OK tell me when I can download it.
Turning the translation into Spanish for the word "Fuel" could be put "Batería restante". If "Batería restante" is too long, just put "Batería" and also match the voice that says "battery xx%".
For "Battery" put "Tensión" that's what we call voltage here.

**Comment by fmartinezo** [comment]
Altitude, Distance and Current can be as I indicated

**Comment by teckel12** [comment]
@fmartinezo The update process will always be the same.  As soon as I push changes to the `horus` branch, you can download them to your transmitter.  The process is:

1. Download [LuaTelemetry-horus.zip](https://github.com/iNavFlight/LuaTelemetry/archive/horus.zip)
1. Open the `LuaTelemetry-horus.zip` file
1. Open the `LuaTelemetry-horus` folder and navigate to the `dist` folder
1. Copy the contents of the `dist` folder (both the `iNav.lua` file and `iNav` folder with contents) to the transmitter's SD card's `\SCRIPTS\TELEMETRY\` folder

**Comment by fmartinezo** [comment]
@teckel12 Is it available now?

**Comment by fmartinezo** [comment]
@teckel12 Hi, I just downloaded the script and checked it. The fps have improved. They have gone from being around 5 (4.9, 5, 5.1) to being around 7 (6.9, 7, 7.1).
If you want something else, let me know and tomorrow I will resume the tests, now it's a bit late for me and I'm going to rest.

**Comment by teckel12** [comment]
@fmartinezo Thanks. The latest release you tried uses images for background and foreground instead of drawing the entire screen.  I figured that would speed things up. 

I got the widget script to seemingly work but It was unstable. I'm not sure if it's just the simulator or not. I'm getting the X10S in a couple days so I'll probably wait to test on an actual transmitter before releasing.

At this point the build you're running should be ready to fly. Fly and let me know what happens.

**Comment by fmartinezo** [comment]
Hello Tim, this weekend if there are no problems with the weather I will try the script and I'll tell you how it goes.
I'm signed up in two inav telegram groups. What do you think if I mention that this development exists in case someone else wants to be a tester?

**Comment by teckel12** [comment]
@fmartinezo Testing is open to anyone, with the knowledge the script is untested in the field. There's safeguards in OpenTX for a script using too many CPU cycles or crashing. And the core of the script is well-tested on Taranis. Only the display code is new for Horus.

With that said, there's some setup to do before using Lua Telemetry, which takes a bit of effort.

But, go ahead and let others know, with the above slight warning.

**Comment by fmartinezo** [comment]
Fantastic news I look forward to the widget and as always ready to do the checkup.
I have already told to the people that the script exists and that it would be good for us to do tests, logically under their own responsibility. I have sent them to this page so they can see how it goes, what has been done and how to download it and put it to work

**Comment by fmartinezo** [comment]
@teckel12 This has been sent by a partner who has tried it. I have not been given the case, since I deactivate the script before dismounting, but when I have occasion I try it and I confirm it. He says:
 
Well, on Saturday I tried it once and it worked well in flight. I landed and when disarming I get on the screen "emergency mode".

The radio only  responded to the control sticks, the switches did not respond ...

**Comment by teckel12** [comment]
@fmartinezo Not sure what emergency mode is. Sounds like it could be crash and the transmitter enters a crippled mode. The error is defiantly not directly from Lua Telemetry as it has no error like that.

I'll look at OpenTX for that error and obviously try to duplicate.

**Comment by fmartinezo** [comment]
Have you received your horus?

**Comment by ZinhoRacing** [comment]
Hi,
I had Emergency mode on the Horus if trying to write text on lcd outside of the lcd area in Lua script.
There is an issu in Open TX for that. If it can help.

@teckel12, I installed LuaTelem script on my horus (with crossifire, I know not working yet). The view is great ! :)
I had a quick read of your source code, I found that file data.lua is handling the reading of telem values. I just wanted to try to make the attitude works to see how it behaves (I thought of changing only the names for pitch, yaw and roll as these channels are available in telemetry in OpenTX over CRSF).
But unfortunatly Companion reporting a nill error when I try to compile ... I used Horus Branch code.
If you have a clue on the problem I would be happy to know, if not, please don't waste your time on it, I would prefer that you can progress on Crossfire integration for sure. :)

**Comment by teckel12** [comment]
> Have you received your horus?

Yes, I received my Horus X10S.  I flashed OpenTX and installed Lua Telemetry, but haven't yet imported models to do real testing.  In the next couple days I'll have a fully-functional Horus test.

**Comment by teckel12** [comment]
@ZinhoRacing Lua Telemetry shouldn't be writing text outside the LCD area, I'm fairly careful about that.  But, there could always be an unforeseen condition that could cause it.  I'll look into it.

For the attitude, it only needs pitch and roll, and Lua Telemetry already looks for two sensors `Ptch` and `Roll` which I believe is what Crossfire already names them.  But, I believe Crossfire outputs these values as radians while INAV outputs them as decimal (multiplied by 10 so it's just an integer).  So, you should be able to just do the following in `iNav.lua` at about lines 128-129 and it *should* work:

```
			data.pitch = math.deg(getValue(data.pitch_id)) * 10
			data.roll = math.deg(getValue(data.roll_id)) * 10
```

BTW, `data.pitchRoll` is set to true if the sensors `Ptch` and `Roll` exist.

**Comment by effysh** [comment]
wonderful job guys where the updated script can be downloaded for testing?
is this is the one: https://github.com/iNavFlight/LuaTelemetry/archive/horus.zip ?

**Comment by ZinhoRacing** [comment]
@teckel12 it is working (for native crsf telemetry data), look nice.

![inkedimg_1377_li](https://user-images.githubusercontent.com/41761785/49766816-b2564800-fcd6-11e8-9fd3-5ac1df481876.jpg)

I like the large pilot view.

**Comment by teckel12** [comment]
@ZinhoRacing Very cool! I'm going to work on my Horus/Crossfire development tomorrow. Unless I get distracted by my EMAX Tiny Hawk BnF I just got TODAY :wink:

**Comment by ZinhoRacing** [comment]
Wowww I want one ! :smiley:
Is it Inav compatible ????  It's a F4 matek F411 look like.

Crazy little thing !! Enjoy.

**Comment by effysh** [comment]
> @effysh Yes, that's the link to download the latest build. Think of it as a nightly build, as it changes nearly every day. Following are the instructions to get the latest latest Horus build. You can do this every day to get the latest.
> 
> 1. Download [LuaTelemetry-horus.zip](https://github.com/iNavFlight/LuaTelemetry/archive/horus.zip)
> 2. Open the `LuaTelemetry-horus.zip` file
> 3. Open the `LuaTelemetry-horus` folder and navigate to the `dist` folder
> 4. Copy the contents of the `dist` folder (both the `iNav.lua` file and `iNav` folder with contents) to the transmitter's SD card's `\SCRIPTS\TELEMETRY\` folder

Hi 
i just tested the script and i get an error when i press to execute the inav lua  file 
all of sensors  are working on my horus 
![20181211_145826_hdr](https://user-images.githubusercontent.com/11709496/49809428-88009b00-fd67-11e8-98dc-33accdd76f53.jpg)
![20181211_150101_hdr](https://user-images.githubusercontent.com/11709496/49810340-b54e4880-fd69-11e8-9181-80eebc2ce5ee.jpg)
![20181211_150120_hdr](https://user-images.githubusercontent.com/11709496/49810480-fba3a780-fd69-11e8-8fb0-6f207ac256d6.jpg)
![20181211_150132_burst01](https://user-images.githubusercontent.com/11709496/49810601-3e657f80-fd6a-11e8-831e-b9a7223180e6.jpg)

**Comment by teckel12** [comment]
@effysh Not sure why it's not working for you. What version of OpenTX are you running? Even if you don't have any sensors it will still run.

**Comment by effysh** [comment]
Ok i just figure it out
I copied the folder dist in to telemetry folder instead of its content (inav folder and inav.lua) now its working 100%
Thanks again for this great script

**Comment by teckel12** [comment]
@effysh Glad you figured it out.  Would different wording for the install instructions make more sense?

**Comment by fmartinezo** [comment]
@teckel12 Hi Tim new spanish voices released. Less metalic and less robotic. Files are on spanish issue (closed) Add spanish support #207

**Comment by teckel12** [comment]
Working on increasing speed and reducing memory.  Also playing with making it a widget.  While I can get the screen to render as a widget, nothing works *yet*.  It has to do with how the script was set up, it just wasn't designed with a widget in mind.  A few more days of tinkering and I hope to have a widget breakthrough.

**Comment by ZinhoRacing** [comment]
Hi @teckel12.
How is the emax Tiny Hawk BnF working ????? 😃

**Comment by effysh** [comment]
Hi i couldnt find any sulution any other place  so i try here ...
 i am using furiusfpv F35 fc
with 1.9.1 firmware with no problem and all telemetry sensors are working and sending values to my horus tx with R9M module and R9 mm receiver
the problems start when i upgrade the F35 fc firmware throw the inav configurator to the latest firmware which is 2.01 and TELEMETRY from F35 stop to work throw S.PORT and only RSSI and RECEIVER BATT works
my opentx is 2.2.2
and the only thing i didn't test is the latest firmware of the R9m and its r9 receiver maybe this is the problem...
any help will be great

**Comment by teckel12** [comment]
> Hi @teckel12.
> How is the emax Tiny Hawk BnF working ????? 😃

Confession bear:  I've been flying it a lot and only working on the Horus/Crossfire stuff when all the batteries are drained.  Going to buy several more batteries.

It flies REALLY well!  So much better than brushed motor indoor quads, much more power.  It's also sturdy and doesn't attract hair and the props don't fly off all the time (yeah, I crash a lot).  Well worth $100.

**Comment by rafaroldan** [comment]
Hi @teckel12 !!! It's working fine in my Horus and a a S800 wing with a Matek 405 and iNav2.0.1
Thanks a lot for the development!!!!

**Comment by teckel12** [comment]
@rafaroldan It seems about done other than possible optimizations which could also happen in other releases.

The only thing I'm still working on is running it as a widget. The script wasn't designed to run as a widget so I'm trying to shoehorn in the script. I believe the only issue is the scope of variables and functions, but it's still been problematic. 

If I don't get somewhere with the widget, I may just release it to be used only as a one-time script. Then, add widget support in a future release. The way the Horus is setup, running as a widget is not really ideal as it kind of takes over everything. And there's been talk of adding more of a Taranis-like support of telemetry scripts to the Horus which would make widget support no longer necessary.

**Comment by Mayolli** [comment]
hi, I just found this discussion.. I like to try and contribute! Downloaded 1.4.3 version (link above is not working for me)
installed and executed but i got ""bad argument to#6 draw line (number ex..." error

**Comment by teckel12** [comment]
@Mayolli 1.4.3 won't work for Horus. 1.4.4 is the Horus release. There's a Wiki about how to get the development build running.

**Comment by Mayolli** [comment]
sorry.. i downloaded wrong file, with one in development all seems to work..
can I contribute with italian translation?

**Comment by rafaroldan** [comment]
Thanks a lot @teckel12! Running as a widget is not mandatory for my but I'm already used to this advantage. I come from Ardupilot world and I run Yaapu FrSky telemetry in two separate telemetry screens as widgets. IMHO widgets in Horus are a neat solution to integrate full telemetry between your model and radio. I also have other normal screens into Horus (with trims, timers, etc..) and you always can scroll through all of them with one button.
Anyhow your work is already outstanding!!! I'm really thankful.

**Comment by fmartinezo** [comment]
@teckel12 Hi tim i can read in the main page that 1.4.4 version is done, but on the downloads page apears only 1.4.3

**Comment by teckel12** [comment]
@fmartinezo Merge conflicts that I'm dealing with.  Sorting this out for a Horus release. When it's sorted out there will be a v1.5.0 under releases.

**Comment by teckel12** [comment]
So...  It's "working" as a widget.  But, button events don't work in widgets **AT ALL**!  It's not a problem I can resolve, as the OpenTX team has decided that the control buttons can't be used with widgets.

So, while it works as a widget, you can't (for example) go to the config menu.  And if you could, you couldn't change anything anyway.  You also can't switch to min/max values.

The "solution" that I'm looking into is to use transmitter controls (which are exposed).  For example, hitting a switch to activate the menu, other switch to go forward/backwards, etc.  Not ideal, and I'm not really sure how well it would work, but looking into it.

Other ideas is to make the config menu required to run as a one-time script or as another widget.  Again, not ideal, but these are the built-in limitations of the widgets.

Maybe someone could pressure the OpenTX team to allow key events to pass through to widgets?

**Comment by effysh** [comment]
@teckel12 awesome work
As far as i understand the config page is used only once so not much of a hassle
Correct me if i am wrong ?

**Comment by teckel12** [comment]
@fmartinezo 1.4.3 is the latest release.  1.5.0 will be the next release, but I'm not quite done with it.  Due to merge conflicts, things are a little messed up, 1.5.0 shows on the main readme right now, but 1.4.3 is the latest release.  The horus link is no longer valid as that branch has been merged into the development and master branches.  

Basically, I know things are a little wonky right now, but everything will be sorted out in the next day.

**Comment by fmartinezo** [comment]
Ok no problem. I'll wait for it

**Comment by fmartinezo** [comment]
@teckel12 Tim step 2 is not clear for me. Copy Horus folder or copy Horus\inav or copy main.lua and main.luac?
it would be like that (1)?
scripts
...........widgets
.................horus
.........................inav
--------------main.lua
--------------main.luac 
or like that (2)?
scripts
...........widgets
.........................inav
--------------main.lua
--------------main.luac 
or like that(3)?
scripts
...........widgets
--------------main.lua
--------------main.luac

**Comment by teckel12** [comment]
@fmartinezo Drag the `iNav` folder in the `Horus` folder to the `\SCRIPTS\WIDGETS` folder on the SD card. I believe that's #2

**Comment by fmartinezo** [comment]
@teckel12 Thanks so much. For the structure that I see I think it can be used as a permanent widget and as a single execution script. Is it like that?

**Comment by teckel12** [comment]
@fmartinezo Correct. It can be setup as a permanent widget or run as a one-time script (just like on the Taranis). The widget just calls the one-time script to keep the same code base.

**Comment by teckel12** [comment]
@fmartinezo This may also be helpful:

https://github.com/iNavFlight/LuaTelemetry/wiki/Installation:-Horus-Widget

**Comment by fmartinezo** [comment]
@teckel12 **Congratulations**, I just tried it, I had a hard time activating the widget but I did it.
There is some error that causes a change of colors or backgrounds on the main screen I explain:
The first time I turn on the radio I have the main screen with two clocks and the photo of the plane, the letters of the clocks come out black.
If I move from the screen to the inav widget screen and return to the main screen, the letters of the clocks go blank

**Comment by teckel12** [comment]
@fmartinezo What user interface theme are you using? I'm using the solid blue one. OpenTX on Horus is kinda stupid in some ways (or maybe not fully thought out). For example, how colors work. 

To insure that text is visible, I set the text color to white. The problem is, the text color is global. So if I set the color, it stays the same for all other widgets. I really need to set the color, or it could be blue text on a blue background.

My guess is that your background color for the template your using is white, and once Lua Telemetry sets the font color to white it's white on white and appears to disappear.

I'm guessing, moew info and maybe a screenshot would be helpful.

**Comment by rafaroldan** [comment]
Thanks again @teckel12, it's working flawlessly as widget in my Horus. Really a good solution using the sticks for the menu... Congrats!
I thinks it is mandatory to use the DarkBlue theme for the widget to work properly. Other telemetry widgets like Yaapu, also require that.

**Comment by teckel12** [comment]
The latest development build adds widget options to activate and set a default font color (for example black if you're using a lighter theme). I only developed/tested with the configurator, but it's a small change. Not sure if the widget options save from session to sessions (would be fairly dumb if they didn't, but it wouldn't surprise me either).

I believe the development install instructions should be accurate, be sure to also update the widget files (obviously).

If you try it out, let me know how it works.

**Comment by teckel12** [comment]
Closed in: https://github.com/iNavFlight/LuaTelemetry/releases/tag/v1.5.0 **Enjoy!**

I fully expect a v1.5.1 release to fix bugs and add enhancements.  Please open new issues for individual items so they can be tracked.
